### DISCUSSION TITLE
How do I use a regular transformer within a program transformer?

### DISCUSSION THREAD

**Discussion by LukasMachetanz** [nonara/ts-patch#29] [Q&A] [answered]
Hey! 

I am wondering of it is possible to get the `TransformationContext` within a `Transformer Program` (https://github.com/nonara/ts-patch#transforming-program)?

Thanks. Greets.

**Comment by nonara** [comment, marked as answer]
Effectively, using a "program transformer" requires a good understanding of the TypeScript Compiler API. 

To help on that journey, I've created a short guide on how to get better acclimated with it. Please have a read here: [Advice for working with the TS Compiler API](https://github.com/nonara/ts-patch/discussions/31)

While I believe that will be helpful for you and others moving forward, I recognize that having a more complete example on how to do this would also be helpful, so I've written the following example:

## Example of transforming a `Progam` with additional AST transformations

```ts
import ts, { CompilerHost, CompilerOptions, Program, SyntaxKind, TransformationContext, SourceFile, Node } from 'typescript';
import { PluginConfig, ProgramTransformerExtras } from "ts-patch";
import path from 'path';
import {} from 'ts-expose-internals'


/* ****************************************************************************************************************** */
// region: Helpers
/* ****************************************************************************************************************** */

/**
 * Patches existing Compiler Host (or creates new one) to allow feeding updated file content from cache
 */
function getPatchedHost(
  maybeHost: CompilerHost | undefined,
  tsInstance: typeof ts,
  compilerOptions: CompilerOptions
): CompilerHost & { fileCache: Map<string, SourceFile> }
{
  const fileCache = new Map();
  const compilerHost = maybeHost ?? tsInstance.createCompilerHost(compilerOptions, true);
  const originalGetSourceFile = compilerHost.getSourceFile;

  return Object.assign(compilerHost, {
    getSourceFile(fileName: string, languageVersion: ts.ScriptTarget) {
      fileName = tsInstance.normalizePath(fileName);
      if (fileCache.has(fileName)) return fileCache.get(fileName);

      const sourceFile = originalGetSourceFile.apply(void 0, Array.from(arguments) as any);
      fileCache.set(fileName, sourceFile);

      return sourceFile;
    },
    fileCache
  });
}

// endregion


/* ****************************************************************************************************************** */
// region: Program Transformer
/* ****************************************************************************************************************** */

export default function transformProgram(
  program: Program,
  host: CompilerHost | undefined,
  config: PluginConfig,
  { ts: tsInstance }: ProgramTransformerExtras,
): Program {
  const compilerOptions = program.getCompilerOptions();
  const compilerHost = getPatchedHost(host, tsInstance, compilerOptions);
  const rootFileNames = program.getRootFileNames().map(tsInstance.normalizePath);

  /* Transform AST */
  const transformedSource = tsInstance.transform(
    /* sourceFiles */ program.getSourceFiles().filter(sourceFile => rootFileNames.includes(sourceFile.fileName)),
    /* transformers */ [ transformAst.bind(tsInstance) ],
    compilerOptions
  ).transformed;

  /* Render modified files and create new SourceFiles for them to use in host's cache */
  const { printFile } = tsInstance.createPrinter();
  for (const sourceFile of transformedSource) {
    const { fileName, languageVersion } = sourceFile;
    const updatedSourceFile = tsInstance.createSourceFile(fileName, printFile(sourceFile), languageVersion);
    compilerHost.fileCache.set(fileName, updatedSourceFile);
  }

  /* Re-create Program instance */
  return tsInstance.createProgram(rootFileNames, compilerOptions, compilerHost);
}

// endregion


/* ****************************************************************************************************************** */
// region: AST Transformer
/* ****************************************************************************************************************** */

/**
 * Change all 'number' keywords to 'string'
 *
 * @example
 * // before
 * type A = number
 *
 * // after
 * type A = string
 */
function transformAst(this: typeof ts, context: TransformationContext) {
  const tsInstance = this;

  /* Transformer Function */
  return (sourceFile: SourceFile) => {
    return tsInstance.visitEachChild(sourceFile, visit, context);

    /* Visitor Function */
    function visit(node: Node): Node {
      if (node.kind === SyntaxKind.NumberKeyword)
        return context.factory.createKeywordTypeNode(tsInstance.SyntaxKind.StringKeyword);
      else
        return tsInstance.visitEachChild(node, visit, context);
    }
  }
}

// endregion
```

**Reply by LukasMachetanz** [reply]
Thanks for providing this example. I really really appreciate it.

**Reply by nonara** [reply]
No problem! Hope it helps.

**Reply by cutephoton** [reply]
@nonara This is interesting, but can you help me understand why one would use this and what the side effects/limitations might be? What is the use cases that cannot be accomplished by other means? It almost seems as if you're loading the files, changing string to number and then performing the full compilation? Is it kind of a pre-processor of sorts?

**Reply by nonara** [reply]
@cutephoton 
> can you help me understand why one would use this and what the side effects/limitations might be? 

Sure. I like the username, by the way!

There are no side effects or limitations outside of the fact that it will add a small amount of additional time to your build process. Generally this is around 1sec or less.

This feature is really only there for certain edge cases which call for it. Here are a couple of examples:

## 1. Writing jsdoc

At one point I needed to modify a source file to add jsdoc comments during regular tsc emit. Unfortunately, this is something that simply was not supported via regular AST manipulation. I'm not sure if this is still an issue, but last I knew, you still couldn't.

I was able to solve this by using a Program transformer. Essentially, I saved the character positions before the start of the functions which needed jsdoc comments. It would then emit the source as strings, modify that text to splice in the comments, and re-create the `SourceFile` objects from the new file text.

The end result handed the compiler valid `SourceFile`s with my added comments

## 2. Updating type info

This is the primary reason that this feature is needed. One limitation of the TS compiler is that if you modify nodes during transformation, you cannot cause the TypeChecker to update its knowledge of the AST's types and symbols. So anything you add or modify essentially has no type information

There are multiple scenarios where that can cause problems. They are a bit too complex to get into here. One real world example is the problem Lukas was facing: https://github.com/nonara/ts-patch/discussions/27

A Program transformer lets you apply transformations and re-walk the types for your modified code before handing the `Program` instance to the compiler.

## 3. Adding / removing source files

This is an example which is easier to understand. You may have a project which generates some of your source code based on config or info in other source files. In that case, you'd want tsc to trigger generation logic first, then update the `Program` instance with the generated files.

An example of all of these being employed can be seen in the `build-tools` logic for [crosstype](https://github.com/crosstype/crosstype), (which I really need to get around to finishing!)

Hope that helps

**Reply by craigphicks** [reply]
Looking at this part 
```
  /* Render modified files and create new SourceFiles for them to use in host's cache */
  const { printFile } = tsInstance.createPrinter();
  for (const sourceFile of transformedSource) {
    const { fileName, languageVersion } = sourceFile;
    const updatedSourceFile = tsInstance.createSourceFile(fileName, printFile(sourceFile), languageVersion);
    compilerHost.fileCache.set(fileName, updatedSourceFile);
  }
```
shouldn't `updatedSourceFile` be the same as `sourceFile`?  I.e., couldn't `compilerHost.fileCache.set(fileName, sourceFile);` have had the same effect?

I am wondering if there used to be JSDoc processing in that part of the loop as you describe in your answer.

**Reply by nonara** [reply]
@craigphicks There are a few key differences in a transformed `SourceFile` and an original. It's been a little bit since I've been in that part of the compiler, so I'll answer to the best of my memory and knowledge.

A transformed `SourceFile` is produced during the emit phase. As a result, it's more of a facsimile or hybrid of a `SourceFile`, as far as the rest of the compiler is concerned. Before the transformation process, I recall that certain properties are stripped, in preparation. Additionally, it lacks position information for nodes which correlate to original source text (as it would, because they no longer correlate). The source text is also never updated. Each updated node contains a sort of chain, going back to the original node.

Most germane to the original intent, I believe that passing a transformed `SourceFile` directly to `Program` will fail to walk the types for the updated source, as there is (partial) existing type data for that file. 

That aside, I believe there are various other ways it will break down, given the earlier issues I mentioned. I imagine it would also be bad from a garbage collection perspective, as there are many references and links associated with the former Program, which you'd like to throw away. These, also, will probably cause issues, as you want everything in a `Program` to be linked to the same. I've experienced a lot of problems related to that sort of thing while investigating routes for this. For my purposes, the right way to go was to produce a set of pristine `SourceFile`, which is what the compiler API is expecting.

You're welcome to test this and look further into it, however. Again, this is from my recollection, and it was a while back, so some of it may be inaccurate. I believe the simple answer is that these stripped / hybrid files aren't meant to be used outside of the emit cycle, and will therefore produce side effects.

If you look into it further and discover anything which confirms or refutes anything I've said, please do share in this thread!

**Reply by zhfnjust** [reply]
this work fine except debugging. i cannot debug my `*.ts` Program after transformed。

When starting the debugger, it always automatically stops at the last line of the generated js file no matter where I set the breakpoint.

<img width="1393" alt="image" src="https://user-images.githubusercontent.com/995909/204141257-0f590ec6-845c-43da-b80f-34b365083239.png">

Here is the minimal codebase to reproduce the problem： 

https://github.com/zhfnjust/plugin-ts

**Reply by bogeeee** [reply]
Had modified you code, cause i forgot to set `"transformProgram": true`.
But [here](https://github.com/bogeeee/program-transformer-example/commit/f18f17549bc8c7984775d907dba32da308980b65)'s a small impovement for that, to give a user friendly error then.

**Reply by bogeeee** [reply]
Got the example working (and thanks for this) and the visitor function replaces a single symbol. But i can't navigate through the AST from there: node.parent is undefined, node.childNodes is undefined (getChildNodes() throws an error), the tsInstance.is...  methods throw an error, ...
Contrary to the [visitor function in your source transformer](https://www.npmjs.com/package/ts-patch#source-transformer-example) example, where this is all working fine. Can you have a look at that ? 
[Here's the repo](https://github.com/bogeeee/program-transformer-example/commit/f18f17549bc8c7984775d907dba32da308980b65) where your example is set up.

**Reply by bogeeee** [reply]
Can confirm that source maps are screwed up in my project as well. Even if the transformer does not insert a single statement.
@zhfnjust: Did you find a solution so far ?
